generators/chipyard/src/main/scala/IOBinders.scala

```diff
@@ -65,42 +67,55 @@ class ComposeIOBinder[T](fn: => (Clock, Bool, Bool, T) => Seq[Any])(implicit tag
 
 // DOC include end: IOBinders
 
-class WithGPIOTiedOff extends OverrideIOBinder({
-  (c, r, s, top: HasPeripheryGPIOModuleImp) => top.gpio.map(gpio => gpio.pins.map(p => p.i.ival := false.B)); Nil
-})
-
-class WithSimBlockDevice extends OverrideIOBinder({
-  (c, r, s, top: CanHavePeripheryBlockDeviceModuleImp) => top.connectSimBlockDevice(c, r); Nil
-})
-
-class WithBlockDeviceModel extends OverrideIOBinder({
-  (c, r, s, top: CanHavePeripheryBlockDeviceModuleImp) => top.connectBlockDeviceModel(); Nil
-})
-
-class WithLoopbackNIC extends OverrideIOBinder({
-  (c, r, s, top: CanHavePeripheryIceNICModuleImp) => top.connectNicLoopback(); Nil
-})
-
-class WithSimNIC extends OverrideIOBinder({
-  (c, r, s, top: CanHavePeripheryIceNICModuleImp) => top.connectSimNetwork(c, r); Nil
-})
-
-class WithUARTAdapter extends OverrideIOBinder({
-  (c, r, s, top: HasPeripheryUARTModuleImp) => {
-    val defaultBaudRate = 115200 // matches sifive-blocks uart baudrate
-    top.uart.zipWithIndex.foreach{ case (dut_io, i) =>
-      val uart_sim = Module(new UARTAdapter(i, defaultBaudRate)(top.p))
-      uart_sim.io.uart.txd := dut_io.txd
-      dut_io.rxd := uart_sim.io.uart.rxd
-    }
+class WithChipTopSimpleClockAndReset extends OverrideIOBinder({
+  (c, r, s, top: HasChipTopSimpleClockAndReset) => {
+    top.clock := c
+    top.reset := r
     Nil
   }
 })
 
+class WithGPIOTiedOff extends Config(new OverrideIOBinder({
+  (c, r, s, top: HasPeripheryGPIOModuleImp) => top.gpio.foreach(_.pins.foreach(_.i.ival := false.B)); Nil
+}) ++ new OverrideIOBinder({
+  (c, r, s, top: CanHaveChipTopGPIO) => top.gpio.foreach(_ := false.B); Nil
+}))
+
+class WithSimBlockDevice extends Config(new OverrideIOBinder({
+  (c, r, s, top: CanHavePeripheryBlockDeviceModuleImp) => SimBlockDevice.connect(c, r, top.bdev)(top.p); Nil
+}) ++ new OverrideIOBinder({
```

Comment:
This paradigm smells fishy to me. Is there a way to do this without duplicating the traits between ChipTop and DigitalTop?

What if instead of calling IOBinders at the Harness level, we call them from within ChipTop? This avoids

Reply:
I thought about something like that, but I don't think it can work. You'll still need to connect things at the TestHarness level, but you'll have no way to match on traits for those IOBinders. I also tried to make `ChipTop` type-parameterized (`[+T <: SystemModule[System]]`) but it ended up being kind of a pain. We could try that route again if desired, but I think it's still going to require the same `IOBinder` duplication unless we scrap support for raw `Top`, which I was trying to avoid.

Reply:
What are the reasons to keep around support for raw `Top`?

Reply:
I haven't looked into any FireSim bridge stuff yet, so I assume that would break at a minimum. But if there's no opposition I do think it would be cleaner to just get rid of raw `Top` support.

Reply:
@davidbiancolin can FireSim bridges be instantiated anywhere in the design? Can we have the module hierarchy be FireSimTestHarness->ChipTop->DigitalTop, and instantiate the bridges within ChipTop?

Reply:
This kinda came up on Monday, but yes we can put the bridges anywhere.